feat: add an example on using autotuning. (#754)
huggingface/kernels: docs/source/builder/triton-autotune.md

# Ship Triton autotune configurations

Triton kernels typically have parameters — tile sizes, number of warps,

> number of pipeline stages — whose optimal values depend on the GPU and the

Reviewer: ```suggestion
Triton kernels typically have parameters, such as tile sizes, number of warps and 
number of pipeline stages, whose optimal values depend on the GPU and the
```

I'm very hesitant about em-dashes these days. Since native speakers told me that the Oxford comma is just a matter of taste, leaving it out makes comma-deliniation work.

Author: Interesting. Some of my native English speaking friends suggested the use of em dashes in academic writing. But I will roll with the suggestion.

problem shape. Autotuning finds good values for these parameters, but doing
it at runtime (e.g. with the
[`@triton.autotune`](https://triton-lang.org/main/python-api/generated/triton.autotune.html)
decorator) re-benchmarks every candidate configuration in each new process.

However, one can run the tuner for the GPU models they like and store the
best found configurations as files for using them later. This effectively
reduces the potentially costly tuning time.

> `kernels` support this by packaging these configurations as JSON files

Reviewer: ```suggestion
`kernels-builder` support this by packaging these configurations as JSON files
```

Since there is no special handling for this in `kernels` itself.

> `kernels` support this by packaging these configurations as JSON files

Author: ```suggestion
`kernel-builder` support this by packaging these configurations as JSON files
```

with the kernel. At runtime, the kernel looks up the configuration for the
current GPU and shape and falls back to sensible defaults when there is no
matching configuration.

This is how, for example, the vLLM fused MoE kernel ships on the Hub: the
[`RedHatAI/moe`](https://huggingface.co/RedHatAI/moe/tree/main/torch-ext/moe/configs)
repository contains a `configs` directory with tuned configurations for many
GPUs. This page walks through a small, complete example of the same pattern:
the [`gemm-triton-autotune`](https://github.com/huggingface/kernels/tree/main/examples/kernels/gemm-triton-autotune)
example kernel, a Triton GEMM published as
[`kernels-test/gemm-triton-autotune`](https://huggingface.co/kernels-test/gemm-triton-autotune).

## Shipping data files with a kernel

Configuration files are plain JSON files inside the kernel's Python package,
in `torch-ext/<package_name>/configs/`. By default, only `py` and `pyi`
files are picked up from the package directory, so add `json` to the

> [`pyext` option](writing-kernels.md#torch-noarch) in `build.toml`:

Reviewer: ```suggestion
Since autotune files are plain JSON files, we can store them anywhere inside the
kernels main Python sources in `torch-ext/<kernel_name>`. For this example,
we will use  `torch-ext/<kernel_name>/configs/`. By default, only `py` and `pyi`
files are picked up from the kenel's Python source directory, so add `json` to the
[`pyext` option](writing-kernels.md#torch-noarch) in `build.toml`:
```

Author: Nice I like it. It adds more specificity.

```
configs/N=4096,K=4096,device_name=NVIDIA_L4.json
configs/N=14336,K=4096,device_name=NVIDIA_L4.json
```

Each file maps an `M` value to the best configuration that the autotuner
found for that `M`:

```json
{
    "1": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 128,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 8,
        "num_warps": 4,
        "num_stages": 3
    },
    "1024": {
        "BLOCK_SIZE_M": 128,
        "BLOCK_SIZE_N": 128,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 8,
        "num_warps": 4,
        "num_stages": 3
    }
}
```

Since the device name is part of the file name, configurations tuned for one
GPU are never applied to another. A configuration that would exceed the
resources of a smaller GPU (e.g. shared memory) is therefore harmless to
ship.

## Looking up configurations at runtime

At kernel launch, the kernel checks whether a configuration file exists for
the current device and shape. If it does, the configuration with the nearest
tuned `M` is used; otherwise the kernel falls back to a conservative default
and logs a warning. The lookup is cached, so the file is read at most once

> per process:

Reviewer: I don't think this is true, since by default `maxsize=128`, so given enough N, K combinations, it will reload. Since the size will be bound by the number of files anyway, I think it's better to use `functools.cache`, which is equivalent to `lru_cache(maxsize=None)`.

```python

> @functools.lru_cache

Reviewer: ```suggestion
@functools.cache
```

def _load_tuned_configs(N: int, K: int) -> Optional[Dict[int, Dict[str, int]]]:
    path = _CONFIGS_DIR / config_file_name(N, K)
    if path.exists():
        logger.info("Using tuned GEMM configurations from %s.", path)
        with open(path) as f:
            return {int(m): config for m, config in json.load(f).items()}
    logger.warning(
        "No tuned GEMM configuration found for this device and shape (%s). "
        "Falling back to heuristic defaults, performance may be suboptimal. "
        "Generate a configuration with `tune_gemm(N=%d, K=%d)`.",
        path.name,
        N,
        K,
    )
    return None

def get_config(M: int, N: int, K: int) -> Dict[str, int]:
    tuned = _load_tuned_configs(N, K)

>     if tuned:

Reviewer: ```suggestion
    if tuned is not None:
```

(truthy values are meh)

Author: Ah yeah! I follow it in the other libs too (`diffusers`, `peft`, etc.).
